packages/shared/lib/utils/http.ts

```diff
@@ -0,0 +1,92 @@
+/* eslint-disable @typescript-eslint/no-unnecessary-type-parameters */
+import { Err, Ok, redactHeaders, redactObjectOrString, redactURL } from '@nangohq/utils';
+
+import { NangoInternalError } from './error.js';
+
+import type { LogContextStateless } from '@nangohq/logs';
+import type { HTTP_METHOD, MessageHTTPRequest, MessageRow } from '@nangohq/types';
+import type { Result } from '@nangohq/utils';
+
+export async function loggedFetch<TBody>(
+    {
+        url,
+        method,
+        headers,
+        body
+    }: {
+        url: URL;
+        method?: HTTP_METHOD;
+        headers?: Record<string, string> | undefined;
+        body?: string | null;
+    },
+    options: {
+        // TODO: maybe put that in logCtx
+        valuesToFilter: string[];
+        logCtx: LogContextStateless;
+        context: MessageRow['context'];
+    }
+): Promise<Result<{ res: Response; body: TBody }>> {
+    const props: RequestInit = {
+        headers: new Headers(headers),
+        method: method || 'GET'
+        // TODO: use agent
+    };
+
+    if (body) {
+        props.body = body;
+    }
+
+    const createdAt = new Date();
+    const requestLog: MessageHTTPRequest = {
+        headers: redactHeaders({ headers: headers }),
+        method: props.method!,
+        url: redactURL({ url: url.href, valuesToFilter: options.valuesToFilter }),
+        body: body ? redactObjectOrString({ data: body, valuesToFilter: options.valuesToFilter }) : undefined
+    };
+    try {
+        const res = await fetch(url, props);
+        let body: TBody;
+        const contentType = res.headers.get('content-type');
+        if (contentType?.includes('application/json')) {
+            body = (await res.json()) as TBody;
+        } else {
+            body = (await res.text()) as TBody;
+        }
+
+        if (res.status >= 300) {
+            void options.logCtx.http(`${props.method} ${url.href}`, {
+                request: requestLog,
+                response: { code: res.status, headers: redactHeaders({ headers: res.headers }) },
+                context: options.context,
+                createdAt,
+                meta: { body }
+            });
+        } else {
+            void options.logCtx.http(`${props.method} ${url.href}`, {
+                request: requestLog,
+                response: { code: res.status, headers: redactHeaders({ headers: res.headers }) },
+                context: options.context,
+                createdAt
+            });
+        }
```

Comment:
Note there are two kinds of error, critical one like Network error and HTTP error, I chose to still return Ok on HTTP Error because it's easier to consume. Let me know if you have suggestion